File: Problems/148_Sort_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if(!head || !head->next)
            return head;
        ListNode* middle = getMiddle(head);
        ListNode* left = sortList(head);
        ListNode* right = sortList(middle);
        return merge(left, right);
    }
    
    ListNode* merge(ListNode* head1, ListNode* head2){
        // dummy head
        ListNode* res = new ListNode(0);
        ListNode* curr = res;
        while(head1 && head2){
            if(head1->val<head2->val){
                curr->next = head1;
                head1 = head1->next;
            }else{
                curr->next = head2;
                head2 = head2->next;
            }
            curr = curr->next;
        }
        if(!head1) curr->next = head2;
        else curr->next = head1;
        return res->next;
    }
    
    ListNode* getMiddle(ListNode* head){
        ListNode* slow = NULL;
        ListNode* fast = head;
        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow==NULL ? head :slow->next;
        }
        ListNode* middle = slow->next;
        slow->next = NULL;
        return middle;
    }
};

File: Problems/148_Sort_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# MergeSort solution
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        def mergeSort(head):
            if not head or not head.next:
                return head
            right = findMiddle(head)
            left = head
            left = mergeSort(left)
            right = mergeSort(right)
            newHead = merge(left, right)
            return newHead

        def findMiddle(head):
            slow = head
            fast = head
            
            while fast.next and fast.next.next:
                fast = fast.next.next
                slow = slow.next
            res = slow.next
            slow.next = None
            return res

        def merge(left, right):
            dummy = ListNode()
            curr = dummy
            while left and right:
                if left.val < right.val:
                    curr.next = left
                    left = left.next
                else:
                    curr.next = right
                    right = right.next
                curr = curr.next

            if not left:
                curr.next = right
            else:
                curr.next = left
            return dummy.next
        
        head = mergeSort(head)
        return head

class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        nums = []
        curr = head
        while curr:
            nums.append(curr.val)
            curr = curr.next
        nums.sort()
        curr = head
        i = 0
        while curr:
            curr.val = nums[i]
            i += 1
            curr = curr.next
        return head